File: common/src/main/java/com/wolfyscript/utilities/gui/SignalImpl.java
package com.wolfyscript.utilities.gui;

import com.wolfyscript.utilities.gui.signal.Signal;

import java.util.HashSet;
import java.util.Objects;
import java.util.Set;
import java.util.function.Function;
import java.util.function.Supplier;

public class SignalImpl<MT> implements Signal<MT> {

    private final String key;
    private final Class<MT> messageValueType;
    private MT value;

    private final Set<Effect> linkedItems = new HashSet<>();

    public SignalImpl(String key, GuiViewManager viewManager, Class<MT> messageValueType, Supplier<MT> defaultValueFunction) {
        this.key = key;
        this.messageValueType = messageValueType;
        this.value = defaultValueFunction.get();
        this.viewManager = viewManager;
    }

    @Override
    public void linkTo(Effect item) { // TODO: Rethink this, move it out of the signal
        linkedItems.add(item);
    }

    @Override
    public String key() {
        return this.key;
    }

    @Override
    public Class<MT> valueType() {
        return messageValueType;
    }

    @Override
    public void set(MT newValue) {
        this.value = newValue;
        ((GuiViewManagerImpl) viewManager).updateObjects(linkedItems);
    }

    @Override
    public void update(Function<MT, MT> function) {
        set(function.apply(value));
    }

    @Override
    public MT get() {
        return value;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        SignalImpl<?> signal = (SignalImpl<?>) o;
        return Objects.equals(key, signal.key) && Objects.equals(messageValueType, signal.messageValueType);
    }

    @Override
    public int hashCode() {
        return Objects.hash(key, messageValueType);
    }

}

File: common/src/main/java/com/wolfyscript/utilities/gui/reactivity/SignalImpl.kt
/*
 *       WolfyUtilities, APIs and Utilities for Minecraft Spigot plugins
 *                      Copyright (C) 2021  WolfyScript
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.wolfyscript.utilities.gui.reactivity

import java.util.*
import java.util.function.Function
import kotlin.reflect.KClass
import kotlin.reflect.safeCast

class SignalImpl<MT : Any>(private val id: NodeId, private val type: KClass<MT>) :
    Signal<MT> {
    private var tagName: String? = null

    override fun tagName(tagName: String) {
        this.tagName = tagName
    }

    override fun tagName(): String {
        return if (tagName == null || tagName!!.isBlank()) ("internal_$id") else tagName!!
    }

    fun id(): NodeId {
        return this.id
    }

    override fun set(newValue: MT) {
        id.runtime.reactiveSource.setValue(id, type, newValue)
    }

    override fun update(function: Function<MT, MT>) {

    }

    override fun get(): MT? {
        val reactivityNode: ReactivityNode<*>? = id.runtime.reactiveSource.untypedNode(id)
        if (type.isInstance(reactivityNode?.value)) {
            return type.safeCast(reactivityNode?.value)
        }
        return null
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val signal = other as SignalImpl<*>
        return id == signal.id
    }

    override fun hashCode(): Int {
        return Objects.hash(id)
    }
}
